Could not build

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: macos, macOS, 11.6, 
CC version: Apple clang version 12.0.5 (clang-1205.0.22.11)

getwd: /Users/username/code/lang/vlang/vlang
vmodules: /Users/username/.vmodules
vroot: /Users/username/src/v
vexe: /Users/username/src/v/v
vexe mtime: 2021-09-21 02:52:24
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 a8905e2.b2ecca3

Git version: git version 2.33.0
Git vroot status: weekly.2021.37-51-gb2ecca39-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 689c8a02

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg study/issue/optional_sum_type_if.v`
```v
struct SA {}
struct SB {}
struct SC {}
type Sum = SA | SB | SC
fn main() {
	m := map[string]Sum
	v := if a := m['a'] {
		println('a => $a')
		return_optional(a) ? // Fails only if the expr in this line has or_block (or_block kind (.propagation or .block) doesn't matter)
	} else {
		2
	}
	println('v => $v')
}

fn return_optional(input Sum) ?int {
	return 1
}
```

**What did you expect to see?**

successfully built


**What did you see instead?**
```
==================
                      ^
/tmp/v/optional_sum_type_if.3017233788482767169.tmp.c:10427:7: error: use of undeclared identifier '_t4'; did you mean '_t2'?
                if (_t4.state != 0) { /*or block*/ 
                    ^~~
                    _t2
/tmp/v/optional_sum_type_if.3017233788482767169.tmp.c:10415:19: note: '_t2' declared here
        Option_main__Sum _t2 = {0};
                         ^
/tmp/v/optional_sum_type_if.3017233788482767169.tmp.c:10428:28: error: use of undeclared identifier '_t4'; did you mean '_t2'?
                        panic_optional_not_set(*_t4.err.msg);
                                                ^~~
                                                _t2
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4880,6 +4880,8 @@ fn (mut g Gen) need_tmp_var_in_if(node ast.IfExpr) bool {
 							if left_sym.kind in [.array, .array_fixed, .map] {
 								return true
 							}
+						} else if stmt.expr.or_block.kind != .absent {
+							return true
 						}
 					}
 				}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -601,9 +601,9 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 		// if !g.is_autofree {
 		g.or_block(tmp_opt, node.or_block, node.return_type)
 		//}
+		unwrapped_typ := node.return_type.clear_flag(.optional)
+		unwrapped_styp := g.typ(unwrapped_typ)
 		if is_gen_or_and_assign_rhs {
-			unwrapped_typ := node.return_type.clear_flag(.optional)
-			unwrapped_styp := g.typ(unwrapped_typ)
 			if unwrapped_typ == ast.void_type {
 				g.write('\n $cur_line')
 			} else if g.table.get_type_symbol(node.return_type).kind == .multi_return {
@@ -616,7 +616,7 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 				}
 			}
 		} else {
-			g.write('$curr_line ${tmp_opt}.data')
+			g.write('$curr_line (*($unwrapped_styp*)${tmp_opt}.data)')
 		}
 	} else if gen_keep_alive {
 		if node.return_type == ast.void_type {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -574,12 +574,14 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 		''
 	}
 	tmp_opt := if gen_or || gen_keep_alive { g.new_tmp_var() } else { '' }
+	mut curr_line := ''
 	if gen_or || gen_keep_alive {
 		mut ret_typ := node.return_type
 		if gen_or {
 			ret_typ = ret_typ.set_flag(.optional)
 		}
 		styp := g.typ(ret_typ)
+		if gen_or && !is_gen_or_and_assign_rhs { curr_line = g.go_before_stmt(0) }
 		g.write('$styp $tmp_opt = ')
 	}
 	if node.is_method && !node.is_field {
@@ -611,6 +613,8 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 					g.write('\n $cur_line $tmp_opt')
 				}
 			}
+		} else {
+			g.write('$curr_line ${tmp_opt}.data')
 		}
 	} else if gen_keep_alive {
 		if node.return_type == ast.void_type {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -563,7 +563,7 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 		&& g.pref.gc_mode in [.boehm_full, .boehm_incr, .boehm_full_opt, .boehm_incr_opt]
 	gen_or := node.or_block.kind != .absent // && !g.is_autofree
 	is_gen_or_and_assign_rhs := gen_or && !g.discard_or_result
-	cur_line := if is_gen_or_and_assign_rhs || gen_keep_alive { // && !g.is_autofree {
+	mut cur_line := if is_gen_or_and_assign_rhs || gen_keep_alive { // && !g.is_autofree {
 		// `x := foo() or { ...}`
 		// cut everything that has been generated to prepend optional variable creation
 		line := g.go_before_stmt(0)
@@ -574,7 +574,6 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 		''
 	}
 	tmp_opt := if gen_or || gen_keep_alive { g.new_tmp_var() } else { '' }
-	mut curr_line := ''
 	if gen_or || gen_keep_alive {
 		mut ret_typ := node.return_type
 		if gen_or {
@@ -582,7 +581,7 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 		}
 		styp := g.typ(ret_typ)
 		if gen_or && !is_gen_or_and_assign_rhs {
-			curr_line = g.go_before_stmt(0)
+			cur_line = g.go_before_stmt(0)
 		}
 		g.write('$styp $tmp_opt = ')
 	}
@@ -603,20 +602,16 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 		//}
 		unwrapped_typ := node.return_type.clear_flag(.optional)
 		unwrapped_styp := g.typ(unwrapped_typ)
-		if is_gen_or_and_assign_rhs {
-			if unwrapped_typ == ast.void_type {
-				g.write('\n $cur_line')
-			} else if g.table.get_type_symbol(node.return_type).kind == .multi_return {
-				g.write('\n $cur_line $tmp_opt /*U*/')
-			} else {
-				if !g.inside_const {
-					g.write('\n $cur_line (*($unwrapped_styp*)${tmp_opt}.data)')
-				} else {
-					g.write('\n $cur_line $tmp_opt')
-				}
-			}
+		if unwrapped_typ == ast.void_type {
+			g.write('\n $cur_line')
+		} else if g.table.get_type_symbol(node.return_type).kind == .multi_return {
+			g.write('\n $cur_line $tmp_opt /*U*/')
 		} else {
-			g.write('$curr_line (*($unwrapped_styp*)${tmp_opt}.data)')
+			if !g.inside_const {
+				g.write('\n $cur_line (*($unwrapped_styp*)${tmp_opt}.data)')
+			} else {
+				g.write('\n $cur_line $tmp_opt')
+			}
 		}
 	} else if gen_keep_alive {
 		if node.return_type == ast.void_type {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -581,7 +581,9 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 			ret_typ = ret_typ.set_flag(.optional)
 		}
 		styp := g.typ(ret_typ)
-		if gen_or && !is_gen_or_and_assign_rhs { curr_line = g.go_before_stmt(0) }
+		if gen_or && !is_gen_or_and_assign_rhs {
+			curr_line = g.go_before_stmt(0)
+		}
 		g.write('$styp $tmp_opt = ')
 	}
 	if node.is_method && !node.is_field {

--- a/vlib/v/tests/if_expression_test.v
+++ b/vlib/v/tests/if_expression_test.v
@@ -228,3 +228,9 @@ fn test_if_expr_with_optional() ? {
 	}
 	assert v == 2
 }
+
+fn test_if_expr_with_or_block() {
+	arr := ['a']
+	a := if arr.len == 0 || arr[0] == '-' { 123 } else { return_optional() or { -1 } }
+	assert a == 1
+}

--- a/vlib/v/tests/if_expression_test.v
+++ b/vlib/v/tests/if_expression_test.v
@@ -213,3 +213,18 @@ fn test_if_expr_with_complex_array_methods() {
 	println(ret)
 	assert ret == ['a', 'b', 'c']
 }
+
+fn return_optional() ?int {
+	return 1
+}
+
+fn test_if_expr_with_optional() ? {
+	m := map[string]int{}
+	v := if a := m['a'] {
+		println('$a')
+		return_optional() ?
+	} else {
+		2
+	}
+	assert v == 2
+}

